ocaml/typing/solver.ml

```diff
@@ -0,0 +1,724 @@
+          match disallow_right mv with
+          | Amode b -> loop (C.join obj a b) mvs xs
+          | Amodevar mv -> loop a (mv :: mvs) xs
+          | Amodejoin (b, mvs') -> loop (C.join obj a b) (mvs' @ mvs) xs)
+    in
+    loop (C.min obj) [] l
+
+  let meet (type a l) obj l =
+    let rec loop :
+        a ->
+        (a, disallowed * allowed) morphvar list ->
+        (a, l * allowed) mode list ->
+        (a, disallowed * allowed) mode =
+     fun a mvs ->
+      if C.le obj a (C.min obj)
+      then fun _ -> Amode (C.min obj)
+      else
+        function
+        | [] -> Amodemeet (a, mvs)
+        | mv :: xs -> (
+          match disallow_left mv with
+          | Amode b -> loop (C.meet obj a b) mvs xs
+          | Amodevar mv -> loop a (mv :: mvs) xs
+          | Amodemeet (b, mvs') -> loop (C.meet obj a b) (mvs' @ mvs) xs)
+    in
+    loop (C.max obj) [] l
+
+  let zap_to_floor_morphvar ~commit obj mv =
+    let log, lower = constrain_mlower_try obj mv in
+    if commit then !append_changes log else undo_changes !log;
+    lower
+
+  let zap_to_floor : type a r. a C.obj -> (a, allowed * r) mode -> a =
+   fun obj -> function
+    | Amode a -> a
+    | Amodevar mv -> zap_to_floor_morphvar ~commit:true obj mv
+    | Amodejoin (a, mvs) ->
+      List.fold_left
+        (fun acc mv ->
+          C.join obj acc (zap_to_floor_morphvar ~commit:true obj mv))
+        a mvs
+
+  (* because lower bound conservative, this check is also conservative.
+     if it returns Some, then definitely a constant.
+     if it returns None, then we don't know anything *)
+  let check_const : type a l r. a C.obj -> (a, l * r) mode -> a option =
+   fun obj -> function
+    | Amode a -> Some a
+    | Amodevar mv ->
+      let lower = zap_to_floor_morphvar ~commit:false obj mv in
+      if C.le obj (mupper obj mv) lower then Some lower else None
+    | Amodemeet (a, mvs) ->
+      let upper =
+        List.fold_left (fun x mv -> C.meet obj x (mupper obj mv)) a mvs
+      in
+      let lower =
+        List.fold_left
+          (fun x mv ->
+            C.meet obj x (zap_to_floor_morphvar ~commit:false obj mv))
+          a mvs
+      in
+      if C.le obj upper lower then Some upper else None
+    | Amodejoin (a, mvs) ->
+      let upper =
+        List.fold_left (fun x mv -> C.join obj x (mupper obj mv)) a mvs
+      in
+      let lower =
+        List.fold_left
+          (fun x mv ->
+            C.join obj x (zap_to_floor_morphvar ~commit:false obj mv))
+          a mvs
+      in
+      if C.le obj upper lower then Some lower else None
+
+  let print :
+      type a l r.
+      ?verbose:bool -> a C.obj -> Format.formatter -> (a, l * r) mode -> unit =
+   fun ?(verbose = true) obj ppf m ->
+    print_raw obj ~verbose ppf
+      (match check_const obj m with None -> m | Some a -> Amode a)
+
+  let newvar_above (type a) (obj : a C.obj) = function
+    | Amode a when C.le obj (C.max obj) a -> Amode a, false
+    | m ->
+      let m' = newvar obj in
```

Comment:
Can't we just construct a variable bounded below by `m`, instead of indirectly creating a vanilla fresh variable and then constraining it? If we can, then we can get rid of `logging:false`, because `newvar_above` and `newvar_below` are the only places where we turn logging off.

Reply:
That's a great point - will do.

For the context, we started with the logging turned out, but later realised that logging caused by newvar_below/above constitutes a large part of allocation, so we turned it off. Of course, the "proper" solution is to create a `var` primitively as you suggested.

Reply:
That's good point - will do.

Reply:
On a second thought: I think if I do that, I will be replicating much of the logic in `submode_try` - does the removal of `logging:false` justify that duplication?

Reply:
I have [inlined](https://github.com/ocaml-flambda/flambda-backend/pull/1760#issuecomment-1921994779) them. As shown in that post, while the benefit is not obvious compared to the whole compilation, there is also no harm. Also, I do like the new code. Therefore, I'm resolving this conversation - feel free to reopen.